File: frontend/uniBond_Frontend/src/types/post.ts
import type { ModerationCheckResponse } from "@/types/moderation";

export type PostCommentUser = {
  id: string;
  firstName: string;
  lastName: string;
  fullName: string;
  avatar: string;
  role: "student" | "lecturer" | "company" | "tech_lead" | "admin";
};

export type PostComment = {
  id: string;
  content: string;
  createdAt: string;
  user: PostCommentUser;
};

export type Post = {
  id: string;
  authorId: string;
  authorName: string;
  authorAvatar: string;
  authorRole: "student" | "lecturer" | "company" | "tech_lead" | "admin";
  content: string;
  mediaType?: "image" | "video";
  mediaUrl?: string;
  likes: number;
  commentsCount: number;
  reposts: number;
  isLikedByUser?: boolean;
  isRepostedByUser?: boolean;
  comments?: PostComment[];
  createdAt: string;
};

export type PostCreateWithModerationResponse = {
  message: string;
  moderation: ModerationCheckResponse;
  post: Post;
};

File: backend/app/schemas/post.py
from pydantic import BaseModel
from typing import List, Optional
from datetime import datetime

from app.schemas.moderation import ModerationCheckResponse

#-- Base model schemas for Post and PostMedia --#
#-- Request schemas --#
class PostMediaBase(BaseModel):
    media_url: str
    media_type: str


class PostCreate(BaseModel):
    content: Optional[str] = None
    media: Optional[List[PostMediaBase]] = []
    
    

#-- Response schemas --#        
class PostMediaResponse(PostMediaBase):
    id: int

    class Config:
        from_attributes = True


class PostUserResponse(BaseModel):
    id: int
    first_name: str
    last_name: str
    role: str
    avatar_path: Optional[str] = None

    class Config:
        from_attributes = True


class PostCommentCreate(BaseModel):
    content: str


class PostCommentResponse(BaseModel):
    id: int
    content: str
    created_at: datetime
    user: PostUserResponse

    class Config:
        from_attributes = True


class PostInteractionToggleResponse(BaseModel):
    status: str
    count: int


class PostResponse(BaseModel):
    id: int
    content: str | None
    created_at: datetime
    user_id: int
    user: PostUserResponse
    media: List[PostMediaResponse] = []
    
    # Intreaction metadata
    likes_count: int = 0
    reposts_count: int = 0
    comments_count: int = 0
    is_liked_by_user: bool = False
    is_reposted_by_user: bool = False
    comments: List[PostCommentResponse] = []

    class Config:
        from_attributes = True


class PostCreateWithModerationResponse(BaseModel):
    message: str
    moderation: ModerationCheckResponse
    post: PostResponse
